balancer/state.go

```diff
@@ -0,0 +1,186 @@
+		return res
+	}
+	resp, err := c.Do(req)
+	if err != nil {
+		res.Err = err.Error()
+		return res
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		res.Err = err.Error()
+		return res
+	}
+	l, err := parsexml(body)
+	if err != nil {
+		res.Err = err.Error()
+		return res
+	}
+	res.Online, res.Listeners = true, l
+	return res
+}
+
+// checker receives relays on in and sends updated relays on out.
+// the execution of checker can be halted by cancelling ctx.
+func checker(ctx context.Context, in, out chan radio.Relay) {
+	var wg sync.WaitGroup
+	c := &http.Client{
+		Timeout: 3 * time.Second,
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case relay, ok := <-in:
+			if !ok {
+				wg.Wait()
+				close(out)
+				return
+			}
+			wg.Add(1)
+			go func(r radio.Relay) {
+				defer wg.Done()
+				out <- health(ctx, c, relay)
+			}(relay)
+		}
+	}
+}
+
+// update checks all relays and sets the current relay to re-direct to.
+// update also accumulates listeners from each relay.
+func (br *Balancer) update(ctx context.Context) {
+	relays, err := br.storage.Relay(ctx).All()
+	if err != nil {
+		if errors.Is(errors.NoRelays, err) {
+			log.Println("balancer: no relays in database")
+			return
+		}
+		log.Println("balancer: error getting relays:", err)
+		return
+	}
+	// we already know that len(relays) != 0, so sends are non-blocking.
+	in := make(chan radio.Relay, len(relays))
+	out := make(chan radio.Relay, len(relays))
+
+	go checker(ctx, in, out)
+
+	for _, relay := range relays {
+		if relay.Disabled {
+			continue
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case in <- relay:
+		}
+	}
+	close(in)
+
+	// we assume the worst relay is the fallback, which has a score of zero.
+	var winner radio.Relay
+	winner.Stream = br.Conf().Balancer.Fallback
+	br.listeners = 0
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case relay, ok := <-out:
+			if !ok {
+				br.c.Set(winner.Stream)
+				return
+			}
+			err := br.storage.Relay(ctx).Update(relay)
+			if err != nil {
+				log.Printf("balancer: error updating relay %s: %s\n", relay.Name, err)
+				continue
+			}
+
+			if !relay.Online || relay.Disabled || relay.Noredir || relay.Max <= 0 {
+				continue
+			}
+
+			br.listeners += relay.Listeners
+			score := relay.Score()
+			if score > winner.Score() {
+				winner = relay
+			}
+		}
+	}
+}
+
+func (br *Balancer) start(ctx context.Context) error {
+	const op errors.Op = "balancer/start"
+
+	go func() {
+		for {
+			select {
+			case <-time.After(5 * time.Second):
+				ctx, _ := context.WithTimeout(ctx, 5*time.Second)
+				br.update(ctx)
+				err := br.manager.UpdateListeners(ctx, br.listeners)
+				if err != nil {
+					log.Printf("balancer: error updating listeners: %s", err)
+				}
+			case <-ctx.Done():
+				br.stop(ctx)
+				return
+			}
+		}
+	}()
+	log.Println("balancer: listening on", br.serv.Addr)
+	if !xerrors.Is(http.ErrServerClosed, br.serv.ListenAndServe()) {
```

Comment:
two calls to `ListenAndServe`? also this should use the valkyrie errors package once it's supported, this is on the error1.13 branch but haven't tested it completely yet